src/unity/toolkits/drawing_recognition/data_preparation.cpp

```diff
@@ -0,0 +1,382 @@
+            float x = P.x;
+            float y = P.y;
+            if (x < min_x) min_x = x;
+            if (x > max_x) max_x = x;
+            if (y < min_y) min_y = y;
+            if (y > max_y) max_y = y;
+        }
+    }
+    // Align the drawing to top-left corner and scale to [0,255]
+    for (size_t i = 0; i < num_strokes; i++) {
+        flex_list &stroke = raw_drawing[i].mutable_get<flex_list>();
+        flex_list new_stroke;
+        for (size_t j = 0; j < stroke.size(); j++) {
+            float new_x;
+            float new_y;
+            flex_dict &point = stroke[j].mutable_get<flex_dict>();
+            make_point(&P, point);
+            if (max_x == min_x) {
+                // vertical straight line
+                new_x = (float)min_x;
+            } else {
+                new_x = ((P.x - min_x) * 255.0) / (max_x - min_x);
+            }
+            if (max_y == min_y) {
+                // horizontal straight line
+                new_y = (float)min_y;
+            } else {
+                new_y = ((P.y - min_y) * 255.0) / (max_y - min_y);
+            }
+            flex_dict new_point;
+            new_point.push_back(std::make_pair("x", new_x));
+            new_point.push_back(std::make_pair("y", new_y));
+            new_stroke.push_back(new_point);
+        }
+        // Apply RDP Line Algorithm
+        flex_list compressed_stroke = ramer_douglas_peucker(new_stroke, 2.0);
+        simplified_drawing.push_back(compressed_stroke);
+    }
+    return simplified_drawing;
+}
+
+void swap_points(Point *P1, Point *P2) {
+    float temp_x, temp_y;
+    if (P1->x > P2->x) {
+        temp_x = P1->x;
+        temp_y = P1->y;
+        P1->x = P2->x;
+        P1->y = P2->y;
+        P2->x = temp_x;
+        P2->y = temp_y;
+    }
+}
+
+bool in_bounds(int x, int y, int dim) {
+    return (x >= 0 && x < dim && y >= 0 && y < dim);
+}
+
+flex_nd_vec paint_stroke(
+    flex_nd_vec bitmap, Point start, Point end, float stroke_width) {
+    swap_points(&start, &end);
+    float slope = (end.y - start.y)/(end.x - start.x);
+    int x1 = (int)(start.x);
+    int y1 = (int)(start.y);
+    int x2 = (int)(end.x);
+    int y2 = (int)(end.y);
+    int pad = (int)(stroke_width/2);
+    size_t dimension = bitmap.shape()[1];
+    if (x1 == x2) {
+        // just paint the y axis
+        int min_y = MIN(y1, y2);
+        int max_y = MAX(y1, y2);
+        for (int y = min_y; y < max_y; y++) {
+            for (int dx = -pad; dx < pad; dx++) {
+                for (int dy = -pad; dy < pad; dy++) {
+                    // figure out how to do this with the flex_list
+                    if (in_bounds(x1+dx, y+dy, dimension)) {
+                        bitmap[(y+dy)*dimension + (x1+dx)] = 1.0;
+                    }
+                }
+            }
+        }
+    } else {
+        for (int x = x1; x < x2; x++) {
+            int y = (int)(slope * (x - x1) + y1);
+            for (int dx = -pad; dx < pad; dx++) {
```

Comment:
I feel like you could factor out a helper function (or maybe just a lambda) that draws a particular (x,y) point (plus padding)